src/executor.ts

```diff
@@ -125,10 +125,32 @@ export class Executor {
    * Execute the root script.
    */
   async execute(): Promise<Result<ServiceMap, Failure[]>> {
-    // TOOD(aomarks) If we have any running services from a previous watch
-    // iteration, we should at this point shut down any of the ones that have
-    // since been deleted from the build graph entirely, or which have become
-    // non-directly-invoked.
+    if (
+      this._previousIterationServices !== undefined &&
+      this._previousIterationServices.size > 0
+    ) {
+      // If any services were removed from the graph entirely, or used to be
+      // directly invoked but are no longer, then stop them now.
+      const currentDirectlyInvokedServices = new Set<ScriptReferenceString>();
+      for (const script of findAllScripts(this._rootConfig)) {
+        if (script.service && script.isDirectlyInvoked) {
+          currentDirectlyInvokedServices.add(scriptReferenceToString(script));
+        }
+      }
+      const stopPromises = [];
+      for (const [key, service] of this._previousIterationServices) {
+        if (!currentDirectlyInvokedServices.has(key)) {
+          const child = service.detach();
+          if (child !== undefined) {
+            child.kill();
+            stopPromises.push(child.completed);
```

Comment:
can `child.completed` reject? If so use `Promise.allSettled()` below.

Reply:
It cannot